File: CPP_Solutions/Knapsack/Subset_Sum/Is_Subset_Sum_tabulation.cpp
#include <iostream>
// #include <cstring>

using namespace std;

// Considering Constraints n<=100 and sum<=1000

bool isSubsetSum(int arr[], int sum, int n){
    int t[n+1][sum+1];
    // Initialisation
    for(int i = 0; i< n+1; i++)
        t[i][0] = true; // Fill First column with true
    for(int j = 1; j < sum + 1; j++)
        t[0][j] = false; // Fill first row except first element with false
        
    // Main iterative code
    for(int i = 1; i< n+1 ; i++)
        for(int j = 1; j< sum + 1; j++)
            t[i][j] = (arr[i-1] <= j) ? ((t[i-1][j - arr[i-1]]) || (t[i-1][j])) : t[i-1][j];
            
    return t[n][sum];
}

int main()
{
    int arr[] = { 3, 34, 4, 12, 5, 2 };
    int sum = 9;
    int n = sizeof(arr) / sizeof(arr[0]);
    cout << isSubsetSum(arr, sum, n);

    return 0;
}

File: Python_Solutions/Knapsack/Subset_Sum/Is_Subset_Sum_tabulation.py
# Given constraints W<= 1002 and n<=102
def isSubsetSum(arr, sum, n):
    t = [[False for j in range(sum+1)] for i in range(n+1)] # initialisation of enitre matrix with false

    # Initialise first column to true
    for i in range(n+1):
        t[i][0] = True
    
    # Initialise first row excluding first element as False(this loop is not necessary as already all elements 
    # of matrix exlcuding the one's in the previous loop is False)
    # for j in range(1, sum+1):
    #     t[0][j] = False

    # Main Iterative code
    for i in range(1, n+1):
        for j in range(1, sum+1):
            t[i][j] = ((t[ i - 1 ][ j - arr[ i - 1 ]]) or (t[i-1][j])) if (arr[i-1] <= j) else t[i-1][j]

    return t[n][sum]

arr = [3, 34, 4, 12, 5, 2]
sum = 9
n = len(arr)
print(isSubsetSum(arr, sum, n))